sdk/core/src/migration/index.ts

```diff
@@ -0,0 +1,162 @@
+
+    return this.auth
+  }
+
+  /**
+   * Copy of the private method of `AffinidiVaultStorageService`
+   * @param credentials
+   * @private
+   */
+  private async encryptCredentials(credentials: any[]): Promise<VaultCredential[]> {
+    const publicKeyBuffer = this.keysService.getOwnPublicKey()
+    const privateKeyBuffer = this.keysService.getOwnPrivateKey()
+    const encryptedCredentials: VaultCredential[] = []
+
+    for (const credential of credentials) {
+      let credentialId = credential?.id
+      if (!isW3cCredential(credential)) {
+        credentialId = credential?.data?.id
+      }
+
+      const credentialIdHash = await this.platformCryptographyTools.computePersonalHash(privateKeyBuffer, credentialId)
+
+      const typeHashes = []
+      if (isW3cCredential(credential)) {
+        for (const credentialType of credential.type) {
+          const typeHash = await this.platformCryptographyTools.computePersonalHash(privateKeyBuffer, credentialType)
+          typeHashes.push(typeHash)
+        }
+      }
+
+      const cyphertext = await this.platformCryptographyTools.encryptByPublicKey(publicKeyBuffer, credential)
+
+      encryptedCredentials.push({
+        credentialId: credentialIdHash,
+        credentialTypes: typeHashes,
+        payload: cyphertext,
+      })
+    }
+
+    return encryptedCredentials
+  }
+
+  /**
+   * Initiate VCs migration on bloom-vault VC READ event.
+   * @param credentials
+   */
+  async runMigration(credentials: any[]): Promise<void> {
+    const encryptedVCs = await this.encryptCredentials(credentials)
+    await this.runMigrationByChunk(encryptedVCs, 100)
+  }
+
+  /**
+   * Divides given encrypted VCs list into chunks and run migration
+   * @param encryptedVCs
+   * @param chunk
+   */
+  async runMigrationByChunk(encryptedVCs: any[], chunk: number): Promise<void> {
+    for (let i = 0, L = encryptedVCs.length; i < L; i += chunk) {
+      const chunkedVCList = encryptedVCs.slice(i, i + chunk)
+      await this.migrateCredentials(chunkedVCList)
+    }
+  }
+
+  /**
+   * Gets migration status for user with given did
+   * @param didInput
+   */
+  async getMigrationStatus(didInput?: string): Promise<{ status: string }> {
+    const did = didInput || this.didAuthAdapter.did
+    const token = await this.getAuth()
+    const url = `${this.baseUrl}/api/v1/migrationStatus?did=${did}`
+    return this.api.execute(
+      'GET',
+      url,
+      {},
+      {
+        Authorization: `Bearer ${token}`,
+      },
+    )
+  }
+
+  /**
+   * Send list of users encrypted VCs stored on `bloom-vault` to the `vault-migration-service` to start migration process to the `affinidi-vault`
+   * @param {Array<vcMigrationList>} vcList
+   */
+  private async migrateCredentials(vcList: VaultCredential[]): Promise<void> {
+    const token = await this.getAuth()
+    const url = `${this.baseUrl}/api/v1/migrateCredentials`
+    return this.api.execute(
```

Comment:
We need to handle failures here, let's at least add one or two retries if the call fails.

On the broader scope how will we handle failed chunks assuming all retries have failed? Is it possible to retry the whole migration process at a later date?

Reply:
@cosminm-affnd it will retry each time user reading VCs if migration not done. So in worst case when some VCs didn't get to the queue we will be sending all bloom VCs to queue again. [Here](https://replika.atlassian.net/browse/FTL-2109) is described migration status schema and with `CredentialsInQueue` we will avoid duplications in queue with our approach.